requests-for-comments/2022-01-27-future-of-person-model.md

```diff
@@ -0,0 +1,108 @@
+# Request for Comments: Future of person model
+
+- **Created:** 2022-01-27
+- **Owner:** @macobo, @marcushyett-ph 
+
+## Problem: Scaling queries horizontally
+
+In posthog, having insights load fast is key for a great experience. However for large projects tracking >10M active persons per month query performance degrades fast.
+
+This can't be solved by scaling horizontally, because a lot of our queries do aggregation on persons - e.g. filtering by user properties, counting daily active persons and calculating funnels. 
+
+To do these queries we need to JOIN `person` and `person_distinct_id` tables. This in clickhouse is really slow (can cause >30x slower queries) and requires these tables to be replicated on every node, setting an upper bound to scalability. Teams with >10M users often run into memory usage problems as well.
+
+This RFC proposes both a technical and product solution to this problem: Store `person_id` and `person_properties` on the main events table.
+
+However, doing so will require breaking changes to the product. These changes are likely to cause some issues for customers, but these have either a negligible impact (affecting around 0.03% of events) or can be mitigated through documentation and education.
+
+## Proposed solution
+
+From a technical standpoint: Store `person_id` and `person_properties` on the main events table.
+
+From a product standpoint storing person_id and person_properties on events requires making the following changes:
+1. After the first session we will no longer merge events prior to login with the main person’s events.
+2. Person properties become immutable: Changing properties applies only on new events and past data is not updated.
+
+### Customer Impact
+
+#### Benefits
```

Comment:
Extending on 5, do we have analysis on how often users are trying to update person properties atm? I agree with the approach but this could be a crippling experience if we disrupt X% of user experiences with updating person properties

Reply:
These are all good points however I'd argue that for the sake of this proposal, they're mostly related to technical trade-offs. Customer impact here meaning end-users who are using the posthog to do the analysis, not the operators.

> the initial events before logging in aren't tied to the same person

This interpretation isn't quite correct and understanding this is key for this proposal. To avoid writing the same topic twice, see the customer impact solution for graph + explanation + this table in my response to a similar question to kunal: https://github.com/PostHog/meta/pull/39#discussion_r794400029

Read through the proposal again (from a product, not technical standpoint) - would this be an acceptable loss for you as a posthog user?

> 2. Increased disk usage - due to redundancy we now need more storage
> 3. More data to copy during events table future migrations due to redundancy in storing everything on events making migrations slower, requiring more disk overhead
> 4. Some use cases now require 2 steps 1) cohort 2) look up most recent value (e.g. most up to date emails for folks who did X in some timeframe in the past)
> 5. If a critical user property (e.g. is_paying_customer) event went missing (network error whatever) user can currently just set it on the person and all queries will have it (there is the drawback that it's now set for all time, but say we don't care), however in the new system when that property is on events we can't add it later (well we could edit all events, but I assume that's not planned - or well could be done via async migrations if the need arises commonly). Not sure if we care...
> Extending on 5, do we have analysis on how often users are trying to update person properties atm? I agree with the approach but this could be a crippling experience if we disrupt X% of user experiences with updating person properties

I'd argue these are implementation concerns - let's figure them them out should we accept the 2 breaking changes in this proposal.

Resolving this thread for the moment since it's discussing out-of-scope issues and to make the discussion easier to follow. Unresolve if you agree!